### PR TITLE
{apache,caddy,nginx}: not "before" ACME certs using DNS validation

### PR DESCRIPTION
**Pull request by ThinkChaos**
## Description of changes

Relax dependency with certs that are validated via DNS challenge since we know the HTTP server is not required for that validation. This allows marking the server's service as depending on the cert.

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by ThinkChaos** [comment]
Pushed another slightly-related commit to change how the web-servers assert they have access to the ACME certs: use the Systemd service user and groups instead of just the `cfg.user` and `cfg.group`.  
That allows having, for instance, a group per cert so they can be handed out individually via `systemd.services.<name>.serviceConfig.SupplementaryGroups`.

**Comment by m1cr0man** [comment]
@ofborg test acme

**Comment by ThinkChaos** [comment]
That test failure took me a while to figure out!  
Basically Nginx and acme-example.test started at the same time. And (un)luckily we hit the race that causes: Nginx loaded the ACME cert (`fullchain.pem`), but the self-signed key (`key.pem`):
```plain
webserver # [  190.411673] acme-example.test-start[3532]: + cp -vp certificates/_.example.test.crt out/fullchain.pem
webserver # [  190.421142] acme-example.test-start[3589]: 'certificates/_.example.test.crt' -> 'out/fullchain.pem'
webserver # [  190.423763] acme-example.test-start[3532]: + cp -vp certificates/_.example.test.key out/key.pem
webserver # [  190.429723] nginx-pre-start[3585]: nginx: [emerg] SSL_CTX_use_PrivateKey("/var/lib/acme/example.test/key.pem") failed (SSL: error:05800074:x509 certificate routines::key values mismatch)
webserver # [  190.432682] nginx-pre-start[3585]: nginx: configuration file /nix/store/ycs2rnjbhzqlm4q9fjr3w2l7cs994n6g-nginx.conf test failed
webserver # [  190.436718] acme-example.test-start[3590]: 'certificates/_.example.test.key' -> 'out/key.pem'
```

I made the web servers start explicitly after the DNS validated certs.  
User can decide if they want a stronger dependency to guarantee the self-signed certs are never used. (I have Nginx require the certs in my config).

File: nixos/tests/common/acme/server/default.nix [hunk 1]
```diff
@@ -54,11 +54,6 @@ let
   testCerts = import ./snakeoil-certs.nix;
   domain = testCerts.domain;
 
-  resolver = let
-    message = "You need to define a resolver for the acme test module.";
```
Comment (m1cr0man) [inline nixos/tests/common/acme/server/default.nix:58]: What was the reason behind removing this?

**Review by m1cr0man — APPROVED** [review]
Thank you for digging into that test failure. 😄 The race between the web servers and the cert renewal has plagued the test suite for quite some time. Hopefully this results in less flakey tests overall.

Approved I just have one question about a change.

File: nixos/tests/common/acme/server/default.nix (hunk 1 above)
Comment (ThinkChaos) [reply nixos/tests/common/acme/server/default.nix:58]: It's unused so has no effect.

**Comment by ThinkChaos** [comment]
Yeah hopefully the race is fully solved now as we have an explicit order for all certs!

File: nixos/tests/common/acme/server/default.nix (hunk 1 above)
Comment (m1cr0man) [reply nixos/tests/common/acme/server/default.nix:58]: Ah, ok then 😅

Comment (ThinkChaos) [reply nixos/tests/common/acme/server/default.nix:58]: You should setup a LSP it'll highlight this stuff :)

**Comment by max-privatevoid** [comment]
This can cause a service ordering cycle. I've noticed this problem with nginx, but I assume the other servers are also affected.

TL;DR: This happens when the (lexically) first certificate uses the `HTTP-01` challenge, and at least one other certificate uses a different challenge.

`security.acme` has the concept of a "leader" certificate. Apparently, account creation in `go-acme` is not a thing that can be done separately from retrieving a certificate. So what happens instead is one certificate (the first one in lexical order) is picked to be responsible for account creation, and all other certificates depend on it. Once the leader is done and the account has been created, all the other certificate requests can run in parallel.

This becomes a problem when `nginx.service` is ordered `Before` the leader certificate (i.e. if the leader certificate requires the web server to be up, such as when it uses the `HTTP-01` challenge) and also ordered `After` a non-leader certificate. The non-leader certificates are always ordered `After` the leader. The startup order would be `acme-nonleader.service` -> `nginx.service` -> `acme-leader.service` -> `acme-account-xxx.target` -> `acme-nonleader.service`.

This does not fail in the test, because the leader certificate there is a wildcard certificate for `example.test`.

**Comment by ThinkChaos** [comment]
Thanks for taking the time to investigate and report the issue!  
I pushed a quick fix at [ThinkChaos:nixpkgs:fix/acme-account-ordering](https://github.com/NixOS/nixpkgs/compare/master...ThinkChaos:nixpkgs:fix/acme-account-ordering) that prioritizes DNS validated certs for leadership. I'll do more testing and open a PR.

**Comment by m1cr0man** [comment]
I've been sharing my thoughts on a more general solution to the service ordering problem in [Matrix](https://matrix.to/#/!MthpOIxqJhTgrMNxDS:nixos.org/$asVZmWOZlOXRrdByS1SJH57PM9_jlVdMtGT1gSnyPRw?via=m1cr0man.com&via=matrix.org&via=nixos.dev). This all links back to #106862 which is over 4 years old 😅 I will be trying to implement some solutions in the next week

**Comment by ThinkChaos** [comment]
That PR looks good to me but it doesn't touch the leader stuff yet so AFAIK we still need my PR for now.

**Comment by ThinkChaos** [comment]
@misuzu thanks for the report, fix PR is https://github.com/NixOS/nixpkgs/pull/356064

File: nixos/modules/services/web-servers/nginx/default.nix [hunk 2]
```diff
@@ -1209,10 +1211,10 @@ in
         '';
       }
     ] ++ map (name: mkCertOwnershipAssertion {
-      inherit (cfg) group user;
       cert = config.security.acme.certs.${name};
       groups = config.users.groups;
-    }) dependentCertNames;
+      services = [ config.systemd.services.nginx ] ++ lib.optional (cfg.enableReload || vhostCertNames != []) config.systemd.services.nginx-config-reload;
```
Comment (hgl) [inline nixos/modules/services/web-servers/nginx/default.nix:1216]: nginx-config-reload service doesn't have serviceConfig.User, resolving in `error: attribute 'User' missing`

Comment (ThinkChaos) [reply nixos/modules/services/web-servers/nginx/default.nix:1216]: Please provide more context (stack trace at the minimum).

I believe it's failing when formatting the assertion message, so workaround would be to fix the access. I can't say how from just this single line though.  
I'll fix the message if it's indeed that.

Comment (m1cr0man) [reply nixos/modules/services/web-servers/nginx/default.nix:1216]: I'd be really surprised if this is caused by a missing attribute, the test suite would have failed. Did the tests run on this PR?

Comment (ThinkChaos) [reply nixos/modules/services/web-servers/nginx/default.nix:1216]: Yes, and I ran them locally.

Is there a test that verifies the assert fails though? And importantly here, that it fails with the expected message?
That's not something I looked into, but I can add it as part of the fix.

Comment (hgl) [reply nixos/modules/services/web-servers/nginx/default.nix:1216]: I don't get this error when the assertions are satifised, only when they are not. I can provide the stack trace later.

Comment (hgl) [reply nixos/modules/services/web-servers/nginx/default.nix:1216]:
```
       … while calling anonymous lambda
         at /nix/store/n1g84klfb0h3bpwyvc59lcy5ca58h36w-source/lib/attrsets.nix:1570:24:
         1569|     let f = attrPath:
         1570|       zipAttrsWith (n: values:
             |                        ^
         1571|         let here = attrPath ++ [n]; in

       … while calling the 'head' builtin
         at /nix/store/n1g84klfb0h3bpwyvc59lcy5ca58h36w-source/lib/attrsets.nix:1574:11:
         1573|         || pred here (elemAt values 1) (head values) then
         1574|           head values
             |           ^
         1575|         else

       … while calling anonymous lambda
         at /nix/store/n1g84klfb0h3bpwyvc59lcy5ca58h36w-source/lib/attrsets.nix:1204:18:
         1203|         mapAttrs
         1204|           (name: value:
             |                  ^
         1205|             if isAttrs value && cond value

       … from call site
         at /nix/store/n1g84klfb0h3bpwyvc59lcy5ca58h36w-source/lib/attrsets.nix:1207:18:
         1206|             then recurse (path ++ [ name ]) value
         1207|             else f (path ++ [ name ]) value);
             |                  ^
         1208|     in

       … while calling anonymous lambda
         at /nix/store/n1g84klfb0h3bpwyvc59lcy5ca58h36w-source/lib/modules.nix:254:72:
          253|           # For definitions that have an associated option
          254|           declaredConfig = mapAttrsRecursiveCond (v: ! isOption v) (_: v: v.value) options;
             |                                                                        ^
          255|

       … while evaluating the attribute 'value'
         at /nix/store/n1g84klfb0h3bpwyvc59lcy5ca58h36w-source/lib/modules.nix:816:9:
          815|     in warnDeprecation opt //
          816|       { value = addErrorContext "while evaluating the option `${showOption loc}':" value;
             |         ^
          817|         inherit (res.defsFinal') highestPrio;

       … while evaluating the option `system.build.toplevel':

       … while evaluating the attribute 'mergedValue'
         at /nix/store/n1g84klfb0h3bpwyvc59lcy5ca58h36w-source/lib/modules.nix:851:5:
          850|     # Type-check the remaining definitions, and merge them. Or throw if no definitions.
          851|     mergedValue =
             |     ^
          852|       if isDefined then

       … while evaluating a branch condition
         at /nix/store/n1g84klfb0h3bpwyvc59lcy5ca58h36w-source/lib/modules.nix:852:7:
          851|     mergedValue =
          852|       if isDefined then
             |       ^
          853|         if all (def: type.check def.value) defsFinal then type.merge loc defsFinal

       … while evaluating the attribute 'values'
         at /nix/store/n1g84klfb0h3bpwyvc59lcy5ca58h36w-source/lib/modules.nix:845:9:
          844|       in {
          845|         values = defs''';
             |         ^
          846|         inherit (defs'') highestPrio;

       … while evaluating a branch condition
         at /nix/store/n1g84klfb0h3bpwyvc59lcy5ca58h36w-source/lib/modules.nix:841:11:
          840|           # Avoid sorting if we don't have to.
          841|           if any (def: def.value._type or "" == "order") defs''.values
             |           ^
          842|           then sortProperties defs''.values

       … while calling the 'any' builtin
         at /nix/store/n1g84klfb0h3bpwyvc59lcy5ca58h36w-source/lib/modules.nix:841:14:
          840|           # Avoid sorting if we don't have to.
          841|           if any (def: def.value._type or "" == "order") defs''.values
             |              ^
          842|           then sortProperties defs''.values

       … while evaluating the attribute 'values'
         at /nix/store/n1g84klfb0h3bpwyvc59lcy5ca58h36w-source/lib/modules.nix:944:7:
          943|     in {
          944|       values = concatMap (def: if getPrio def == highestPrio then [(strip def)] else []) defs;
             |       ^
          945|       inherit highestPrio;

       … while calling the 'concatMap' builtin
         at /nix/store/n1g84klfb0h3bpwyvc59lcy5ca58h36w-source/lib/modules.nix:944:16:
          943|     in {
          944|       values = concatMap (def: if getPrio def == highestPrio then [(strip def)] else []) defs;
             |                ^
          945|       inherit highestPrio;

       … while calling the 'concatMap' builtin
         at /nix/store/n1g84klfb0h3bpwyvc59lcy5ca58h36w-source/lib/modules.nix:831:17:
          830|         # Process mkMerge and mkIf properties.
          831|         defs' = concatMap (m:
             |                 ^
          832|           map (value: { inherit (m) file; inherit value; }) (addErrorContext "while evaluating definitions from `${m.file}':" (dischargeProperties m.value))

       … while calling anonymous lambda
         at /nix/store/n1g84klfb0h3bpwyvc59lcy5ca58h36w-source/lib/modules.nix:831:28:
          830|         # Process mkMerge and mkIf properties.
          831|         defs' = concatMap (m:
             |                            ^
          832|           map (value: { inherit (m) file; inherit value; }) (addErrorContext "while evaluating definitions from `${m.file}':" (dischargeProperties m.value))

       … while calling the 'map' builtin
         at /nix/store/n1g84klfb0h3bpwyvc59lcy5ca58h36w-source/lib/modules.nix:832:11:
          831|         defs' = concatMap (m:
          832|           map (value: { inherit (m) file; inherit value; }) (addErrorContext "while evaluating definitions from `${m.file}':" (dischargeProperties m.value))
             |           ^
          833|         ) defs;

       … while evaluating definitions from `/nix/store/n1g84klfb0h3bpwyvc59lcy5ca58h36w-source/nixos/modules/system/activation/top-level.nix':

       … from call site
         at /nix/store/n1g84klfb0h3bpwyvc59lcy5ca58h36w-source/lib/modules.nix:832:128:
          831|         defs' = concatMap (m:
          832|           map (value: { inherit (m) file; inherit value; }) (addErrorContext "while evaluating definitions from `${m.file}':" (dischargeProperties m.value))
             |                                                                                                                                ^
          833|         ) defs;

       … while calling 'dischargeProperties'
         at /nix/store/n1g84klfb0h3bpwyvc59lcy5ca58h36w-source/lib/modules.nix:903:25:
          902|   */
          903|   dischargeProperties = def:
             |                         ^
          904|     if def._type or "" == "merge" then

       … while evaluating a branch condition
         at /nix/store/n1g84klfb0h3bpwyvc59lcy5ca58h36w-source/lib/modules.nix:904:5:
          903|   dischargeProperties = def:
          904|     if def._type or "" == "merge" then
             |     ^
          905|       concatMap dischargeProperties def.contents

       … while evaluating the attribute 'value'
         at /nix/store/n1g84klfb0h3bpwyvc59lcy5ca58h36w-source/lib/modules.nix:614:53:
          613|                 (n: value:
          614|                   [{ inherit (module) file; inherit value; }]
             |                                                     ^
          615|                 )

       … while calling the 'throw' builtin
         at /nix/store/n1g84klfb0h3bpwyvc59lcy5ca58h36w-source/nixos/modules/system/activation/top-level.nix:73:10:
           72|   baseSystemAssertWarn = if failedAssertions != []
           73|     then throw "\nFailed assertions:\n${concatStringsSep "\n" (map (x: "- ${x}") failedAssertions)}"
             |          ^
           74|     else showWarnings config.warnings baseSystem;

       … while calling the 'concatStringsSep' builtin
         at /nix/store/n1g84klfb0h3bpwyvc59lcy5ca58h36w-source/nixos/modules/system/activation/top-level.nix:73:41:
           72|   baseSystemAssertWarn = if failedAssertions != []
           73|     then throw "\nFailed assertions:\n${concatStringsSep "\n" (map (x: "- ${x}") failedAssertions)}"
             |                                         ^
           74|     else showWarnings config.warnings baseSystem;

       … while calling anonymous lambda
         at /nix/store/n1g84klfb0h3bpwyvc59lcy5ca58h36w-source/nixos/modules/system/activation/top-level.nix:73:69:
           72|   baseSystemAssertWarn = if failedAssertions != []
           73|     then throw "\nFailed assertions:\n${concatStringsSep "\n" (map (x: "- ${x}") failedAssertions)}"
             |                                                                     ^
           74|     else showWarnings config.warnings baseSystem;

       … from call site
         at /nix/store/n1g84klfb0h3bpwyvc59lcy5ca58h36w-source/nixos/modules/system/activation/top-level.nix:73:77:
           72|   baseSystemAssertWarn = if failedAssertions != []
           73|     then throw "\nFailed assertions:\n${concatStringsSep "\n" (map (x: "- ${x}") failedAssertions)}"
             |                                                                             ^
           74|     else showWarnings config.warnings baseSystem;

       … while calling anonymous lambda
         at /nix/store/n1g84klfb0h3bpwyvc59lcy5ca58h36w-source/nixos/modules/system/activation/top-level.nix:70:27:
           69|
           70|   failedAssertions = map (x: x.message) (filter (x: !x.assertion) config.assertions);
             |                           ^
           71|

       … while evaluating the attribute 'message'
         at /nix/store/n1g84klfb0h3bpwyvc59lcy5ca58h36w-source/nixos/modules/security/acme/mk-cert-ownership-assertion.nix:18:3:
           17|
           18|   message = "Certificate ${cert.domain} (group=${cert.group}) must be readable by service(s) ${
             |   ^
           19|     catSep ", " (map (svc: "${svc.name} (user=${svc.serviceConfig.User} groups=${catSep " " (svcGroups svc)})") services)

       … while calling the 'concatStringsSep' builtin
         at /nix/store/n1g84klfb0h3bpwyvc59lcy5ca58h36w-source/nixos/modules/security/acme/mk-cert-ownership-assertion.nix:19:5:
           18|   message = "Certificate ${cert.domain} (group=${cert.group}) must be readable by service(s) ${
           19|     catSep ", " (map (svc: "${svc.name} (user=${svc.serviceConfig.User} groups=${catSep " " (svcGroups svc)})") services)
             |     ^
           20|   }";

       … while calling anonymous lambda
         at /nix/store/n1g84klfb0h3bpwyvc59lcy5ca58h36w-source/nixos/modules/security/acme/mk-cert-ownership-assertion.nix:19:23:
           18|   message = "Certificate ${cert.domain} (group=${cert.group}) must be readable by service(s) ${
           19|     catSep ", " (map (svc: "${svc.name} (user=${svc.serviceConfig.User} groups=${catSep " " (svcGroups svc)})") services)
             |                       ^
           20|   }";

       error: attribute 'User' missing
       at /nix/store/n1g84klfb0h3bpwyvc59lcy5ca58h36w-source/nixos/modules/security/acme/mk-cert-ownership-assertion.nix:19:49:
           18|   message = "Certificate ${cert.domain} (group=${cert.group}) must be readable by service(s) ${
           19|     catSep ", " (map (svc: "${svc.name} (user=${svc.serviceConfig.User} groups=${catSep " " (svcGroups svc)})") services)
             |                                                 ^
           20|   }";
```

Comment (ThinkChaos) [reply nixos/modules/services/web-servers/nginx/default.nix:1216]: Great thank you for the confirmation!

Comment (magneticflux-) [reply nixos/modules/services/web-servers/nginx/default.nix:1216]: Just ran into this as well. It's caused by `systemd.services.nginx-config-reload` not having a `serviceConfig.User` set here: https://github.com/NixOS/nixpkgs/blob/43f6a895e5169a1a5d35a7141e016ccfc828a28f/nixos/modules/services/web-servers/nginx/default.nix#L1322-L1343

The assertion failed in the first place because nginx didn't set the cert's group automatically. Normally, nginx sets the `group` of the certs it uses to its own `cfg.group`, but this only happens for certs "fully-managed" by nginx: https://github.com/NixOS/nixpkgs/blob/23e89b7da85c3640bbc2173fe04f4bd114342367/nixos/modules/services/web-servers/nginx/default.nix#L1357-L1358
The alternative (in order to use a single wildcard cert) is to set `virtualHosts.*.useACMEHost = "example.com";` instead, but that _doesn't_ trigger this update. To fix this and just do what nginx does to other certs, set `security.acme.certs."example.com".group = config.services.nginx.group;`.

Comment (m1cr0man) [reply nixos/modules/services/web-servers/nginx/default.nix:1216]: @ThinkChaos Taking a step back from the assertion/attribute error itself, what is the value in applying mkCertOwnershipAssertion to nginx-config-reload.service? It does not interact with the certs explicitly, only nginx itself does.

File: nixos/modules/services/web-servers/apache-httpd/default.nix [hunk 3]
```diff
@@ -641,10 +643,10 @@ in
         '';
       }
     ] ++ map (name: mkCertOwnershipAssertion {
-      inherit (cfg) group user;
       cert = config.security.acme.certs.${name};
       groups = config.users.groups;
-    }) dependentCertNames;
+      services = [ config.systemd.services.httpd ] ++ lib.optional (vhostCertNames != []) config.systemd.services.httpd-config-reload;
```
Comment (m1cr0man) [inline nixos/modules/services/web-servers/apache-httpd/default.nix:648]: httpd-config-reload only needs permission for the certs because we do the config test in [ExecStartPre](https://github.com/NixOS/nixpkgs/pull/336412/files#diff-c4e76be6d8d687a02103da358697dd78a98738f3a1bc20c52b3a00fbccf2e6a9R813).

What we could do instead is change the ExecReload entry for httpd itself to run the config test followed by the actual reload, by providing a list of commands (Which get expanded into multiple ExecReload statements). Then we can remove the change on this line.

**Review by m1cr0man — COMMENTED** [review]
A recommendation on how to resolve the assertion problem.

File: nixos/modules/services/web-servers/apache-httpd/default.nix (hunk 3 above)
Comment (ThinkChaos) [reply nixos/modules/services/web-servers/apache-httpd/default.nix:648]: Yes that seems good to me. I didn't want to change those kinds of things in this PR.  
I'll open a fix PR today.

I think we can even remove the reload services altogether but that's for later.

Comment (m1cr0man) [reply nixos/modules/services/web-servers/apache-httpd/default.nix:648]: I made a quick stab at this this evening, and ran the acme test suite against it. Feel free to cherry-pick and open a PR :) https://github.com/NixOS/nixpkgs/compare/master...m1cr0man:nixpkgs:acme-assertion-fix

Comment (ThinkChaos) [reply nixos/modules/services/web-servers/apache-httpd/default.nix:648]: Thanks! I also have a local change to fix the assert code itself.  
Unfortunately, there's no easy way to test asserts since the NixOS test config must build to be ran as a VM. I did play with a workaround, but not sure if it's worth doing that now. I'll think about it more and make a PR later.

Comment (ThinkChaos) [reply nixos/modules/services/web-servers/apache-httpd/default.nix:648]: I opened a simple PR to fix just the assert message: https://github.com/NixOS/nixpkgs/pull/362271

I didn't take your commit, instead made a PR to go a step further and remove the `*-config-reload` services: https://github.com/NixOS/nixpkgs/pull/362296
